File: backend-hono/src/controllers/menuCategoryController.ts
import { type Context } from "hono";
import {
  BaseController,
  controller,
  route,
  httpGet,
  httpPost,
  httpPut,
  httpDelete,
  fromBody,
  producesResponseType,
  authorize,
  fromRoute
} from "#/framework/mvc";
import type { IMenuCategoryService } from "#/core/services/menuCategoryService";
import type { IMenuCategoryApi } from "@hc-management/shared/api";
import {
  MenuCategoryUpsertRequestDto as UpsertRequestDto,
  type MenuCategoryBasicResponseDto as BasicResponseDto
} from "@hc-management/shared/dtos";

@controller
@route("/api/menu-categories")
@authorize
export class MenuCategoryController extends BaseController implements IMenuCategoryApi {
  private readonly menuCategoryService: IMenuCategoryService;

  public constructor(httpContext: Context) {
    super(httpContext);
    this.menuCategoryService = this.serviceProvider.getRequiredService("menuCategoryService");
  }

  @httpGet("/")
  @producesResponseType<BasicResponseDto[]>(200)
  public async getAllAsync(): Promise<BasicResponseDto[]> {
    return this.ok(await this.menuCategoryService.getAllAsync());
  }

  @httpGet("/:id{[0-9]+}")
  @fromRoute(0, "id", "number")
  @producesResponseType<BasicResponseDto>(200)
  @producesResponseType(401)
  @producesResponseType(403)
  @producesResponseType(404)
  public async getSingleAsync(id: number): Promise<BasicResponseDto> {
    return this.ok(await this.menuCategoryService.getSingleAsync(id));
  }

  @httpPost("/")
  @fromBody(0, UpsertRequestDto)
  @producesResponseType<number>(201)
  @producesResponseType(400)
  @producesResponseType(401)
  @producesResponseType(403)
  @producesResponseType(422)
  public async createAsync(requestDto: UpsertRequestDto): Promise<number> {
    const createdId = await this.menuCategoryService.createAsync(requestDto);
    return this.created(`/api/menu-category/${createdId}`, createdId);
  }

  @httpPut("/:id{[0-9]+}")
  @fromRoute(0, "id", "number")
  @fromBody(1, UpsertRequestDto)
  @producesResponseType(200)
  @producesResponseType(400)
  @producesResponseType(401)
  @producesResponseType(403)
  @producesResponseType(404)
  @producesResponseType(422)
  public async updateAsync(id: number, requestDto: UpsertRequestDto): Promise<void> {
    await this.menuCategoryService.updateAsync(id, requestDto);
    return this.ok();
  }

  @httpDelete("/:id{[0-9]+}")
  @fromRoute(0, "id", "number")
  @producesResponseType(200)
  @producesResponseType(401)
  @producesResponseType(403)
  @producesResponseType(404)
  @producesResponseType(422)
  public async deleteAsync(id: number): Promise<void> {
    await this.menuCategoryService.deleteAsync(id);
    return this.ok();
  }
}

File: backend/Api/Controllers/MenuCategoryController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using HCManagement.Core.Features.MenuCategories;

namespace HCManagement.Api.Controllers;

[ApiController]
[Route("api/menu-categories")]
[Authorize]
public class MenuCategoryController : ControllerBase
{
    #region Fields
    private readonly IMenuCategoryService _service;
    #endregion

    #region Constructors
    public MenuCategoryController(IMenuCategoryService service)
    {
        _service = service;
    }
    #endregion

    #region Methods
    [HttpGet]
    [ProducesResponseType<List<MenuCategoryBasicResponseDto>>(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    public async Task<IActionResult> GetAll()
    {
        return Ok(await _service.GetAllAsync());
    }

    [HttpGet("{id:int}")]
    [ProducesResponseType<MenuCategoryBasicResponseDto>(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> GetSingle([FromRoute] int id)
    {
        return Ok(await _service.GetSingleAsync(id));
    }

    [HttpPost]
    [ProducesResponseType<int>(StatusCodes.Status201Created)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(StatusCodes.Status409Conflict)]
    [ProducesResponseType(StatusCodes.Status422UnprocessableEntity)]
    public async Task<IActionResult> Create([FromBody] MenuCategoryUpsertRequestDto requestDto)
    {
        int id = await _service.CreateAsync(requestDto);
        return CreatedAtAction(nameof(GetSingle), new { id }, id);
    }

    [HttpPut("{id:int}")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(StatusCodes.Status409Conflict)]
    [ProducesResponseType(StatusCodes.Status422UnprocessableEntity)]
    public async Task<IActionResult> Update([FromRoute] int id, [FromBody] MenuCategoryUpsertRequestDto requestDto)
    {
        await _service.UpdateAsync(id, requestDto);
        return Ok();
    }

    [HttpDelete("{id:int}")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    [ProducesResponseType(StatusCodes.Status409Conflict)]
    [ProducesResponseType(StatusCodes.Status422UnprocessableEntity)]
    public async Task<IActionResult> Delete([FromRoute] int id)
    {
        await _service.DeleteAsync(id);
        return Ok();
    }
    #endregion
}
 